core/cache/lxdprofilewatcher.go

```diff
@@ -192,6 +194,60 @@ func (w *MachineLXDProfileWatcher) applicationCharmURLChange(_ string, value int
 	logger.Tracef("end of application charm url change %#v", w.applications)
 }
 
+// charmMetadataChange sends a notification if what is saved for its
+// charm lxdprofile changes.  No notification is sent if the profile pointer
+// begins and ends as nil.
+func (w *MachineLXDProfileWatcher) charmMetadataChange(_ string, value interface{}) {
+	// We don't want to respond to any events until we have been fully initialized.
+	select {
+	case <-w.initialized:
+	case <-w.tomb.Dying():
+		return
+	}
+	var notify bool
+	defer func(notify *bool) {
+		if *notify {
+			w.notify()
+			w.metrics.LXDProfileChangeNotification.Inc()
+		} else {
+			w.metrics.LXDProfileNoChange.Inc()
+		}
+	}(&notify)
+
+	charmChange, ok := value.(CharmChange)
+	if !ok {
+		w.logError("programming error, value not of type CharmChange")
+		return
+	}
+
+	// Scan the list of applications assigned to the machine and only
+	// consider the ones whose charm URL matches the one we observed.
+	//
+	// In a typical scenario, the number of applications associated with
+	// a machine will be small so we can simply iterate the list.
```

Comment:
Even if the application list is massive, this is still a cheap operation, as everything is in memory.

Reply:
Yes, I thought that the overhead of maintaining a charmURL -> []appInfos would be overkill for this case.